pubsub/streaming-analytics/PubSubToGCS.py

```diff
@@ -14,119 +14,126 @@
-        }
+    def process(self, element, publish_time=DoFn.TimestampParam):
+        """Processes each windowed element by extracting the message body and its
+        publish time into a tuple and adding a random shard ID as key.
+        """
+        yield (
+            random.randint(0, self.num_shards - 1),
+            (
+                element.decode("utf-8"),
+                datetime.utcfromtimestamp(float(publish_time)).strftime(
+                    "%Y-%m-%d %H:%M:%S.%f"
+                ),
+            ),
+        )
 
 
-class WriteBatchesToGCS(beam.DoFn):
+class WriteToGCS(DoFn):
     def __init__(self, output_path):
         self.output_path = output_path
 
-    def process(self, batch, window=beam.DoFn.WindowParam):
-        """Write one batch per file to a Google Cloud Storage bucket. """
+    def process(self, batch, window=DoFn.WindowParam):
+        """Write messages in a batch to Google Cloud Storage."""
 
         ts_format = "%H:%M"
         window_start = window.start.to_utc_datetime().strftime(ts_format)
         window_end = window.end.to_utc_datetime().strftime(ts_format)
-        filename = "-".join([self.output_path, window_start, window_end])
+        shard_id = str(batch[0])
+        filename = "-".join([self.output_path, window_start, window_end, shard_id])
 
-        with beam.io.gcp.gcsio.GcsIO().open(filename=filename, mode="w") as f:
-            for element in batch:
-                f.write("{}\n".format(json.dumps(element)).encode("utf-8"))
+        with io.gcsio.GcsIO().open(filename=filename, mode="w") as f:
+            for message in batch[1]:
+                f.write(f"{message[0]},{message[1]}\n".encode("utf-8"))
 
 
-def run(input_topic, output_path, window_size=1.0, pipeline_args=None):
-    # `save_main_session` is set to true because some DoFn's rely on
-    # globally imported modules.
+def run(input_topic, output_path, window_size=1.0, num_shards=1, pipeline_args=None):
+    # Set `save_main_session` to True so DoFns can access globally imported modules.
     pipeline_options = PipelineOptions(
         pipeline_args, streaming=True, save_main_session=True
     )
 
-    with beam.Pipeline(options=pipeline_options) as pipeline:
+    with Pipeline(options=pipeline_options) as pipeline:
         (
             pipeline
-            | "Read PubSub Messages"
-            >> beam.io.ReadFromPubSub(topic=input_topic)
-            | "Window into" >> GroupWindowsIntoBatches(window_size)
-            | "Write to GCS" >> beam.ParDo(WriteBatchesToGCS(output_path))
+            # Because `timestamp_attribute` is unspecified in `ReadFromPubSub`, Beam
+            # binds the publish time returned by the Pub/Sub server for each message
+            # to the element. Element timestamp is accessible via `DoFn.TimestampParam`.
+            | "Read from Pub/Sub" >> io.ReadFromPubSub(topic=input_topic)
+            | "Window into" >> GroupMessagesByFixedWindows(window_size, num_shards)
```

Comment:
Totally optional, but it might be easier to read if you inlined the `GroupMessagesByFixedWindows` transform here. That way you can see the whole pipeline from beginning to end without having to scroll up and down to see what's happening.

I would recommend having a custom PTransform like this if you were using it in multiple places so you don't duplicate code. Or if you add unit tests for your transforms, but since we're already doing an e2e test, I would go for readability.

Reply:
I started doing this, then I realized that it would diverge from the Java sample and a screenshot we show on c.g.c. I will keep this in mind when doing the new Pub/Sub Lite to GCS quickstart.